akka-http2-support/src/main/scala/akka/http/impl/engine/http2/AlpnSwitch.scala

```diff
@@ -89,28 +100,59 @@ object AlpnSwitch {
                   propagatePull
                 case Some(ele) ⇒
                   new OutHandler {
-                    def onPull(): Unit = {
+                    override def onPull(): Unit = {
                       out.push(initialElement.get)
                       out.setHandler(propagatePull)
                     }
+
+                    override def onDownstreamFinish(): Unit = {
+                      out.complete()
+                      super.onDownstreamFinish()
+                    }
                   }
                 case None ⇒ propagatePull
               }
 
             out.setHandler(firstHandler)
             setHandler(in, new InHandler {
-              def onPush(): Unit = out.push(grab(in))
+              override def onPush(): Unit = out.push(grab(in))
+
+              override def onUpstreamFinish(): Unit = {
+                out.complete()
```

Comment:
hmm, AFAIR this or the cancel below - was checking manually and pinpointed exactly which one. 

I thought I commented on github which one it was (likely have it noted some where if we want to include just that single fix)